event_model/__init__.py

```diff
@@ -54,7 +56,42 @@ class DocumentRouter:
     Finally, the call to ``router(name, doc)`` returns::
 
         (name, getattr(router, name)(doc))
+
+    Parameters
+    ----------
+    emit: callable, optional
+        Expected signature ``f(name, doc)``
     """
+    def __init__(self, *, emit=None):
+        # Put in some extra effort to validate `emit` carefully, because if
+        # this is used incorrectly the resultant errors can be confusing.
+        if emit is not None:
+            if not callable(emit):
+                raise ValueError("emit must be a callable")
+            sig = inspect.signature(emit)
+            try:
+                # Does this function accept two positional arguments?
+                sig.bind(None, None)
+            except TypeError:
+                raise ValueError("emit must accept two position arguments, name and doc")
+            # Stash a weak reference to `emit`.
+            if inspect.ismethod(emit):
+                self._downstream = weakref.WeakMethod(emit)
```

Comment:
Could "self._downstream" be something like "self._downstream_router" to give some extra context? Or instead include "emit" or "emitter"?